### ISSUE TITLE
Deselect All Filtered not deselecting

### ISSUE THREAD

**Issue by jamespero** [ag-grid/ag-grid#2579]
```
[x] bug report => see 'Providing a Reproducible Scenario'
[] feature request => do not use Github for feature requests, see 'Customers of ag-Grid'
[x] support request => see 'Requesting Community Support'
```
I have a grid with a dedicated checkbox column and filtering available on many columns. I also have checkboxSelection, headerCheckboxSelection, and headerCheckboxSelectionFilteredOnly set to true.

Update:
We also have groups turned on for our use case, and when filtering is applied it's as if the group still thinks the filtered out rows are still present, thus when selecting all rows via the grid header select all, the filtered out rows are not selected thus setting the group's selection to an indeterminate state in turn setting the grid's selection to an indeterminate state.

**Current behavior**

When the grid has any filtering applied, clicking the select/deselect all checkbox, all visible rows are selected, but never deselected.

When the grid does not have any filtering applied select/deselect all works as expected.

**Expected behavior**

When the grid has filtering applied, clicking deselect all checkbox should deselect all selected rows.

**Please tell us about your environment:**

Windows 10, 
VS Code, 
NPM, 
Webpack, 
Angular 5.0.3

* **ag-Grid version:** X.X.X

17.0.0 Angular, Enterprise

* **Browser:**

Chrome, IE11, Firefox, Safari

* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
TypeScript

**Comment by jamespero** [comment]
Update: When setting headerCheckboxSelectionFilteredOnly to false or removing it, select/deselect all works again, but as expected for all rows and not just the filtered ones.

**Comment by jamespero** [comment]
Update: I've found that the select/deselect all checkbox is never switching to the all selected state when I have filtered rows. 

When the rows are filtered and I select all the select all checkbox is marked with the indeterminate state.

**Comment by jamespero** [comment]
Update:

It seems our filtering is returning the full data set in api.filterManager.rowModel.rowsToDisplay where as the plunker example in the documentation returns only the filtered rows. 

Still investigating.

**Comment by jamespero** [comment]
Ok I'm able to reproduce this behavior in a plunker example:

[https://plnkr.co/edit/2hB3XYrjhUhN9cnm4OHo?p=preview](url)

1. Filter the Athletes column by only Michael Phelps
2. Expand the United State group
3. Click the header checkbox to select all
4. Notice header checkbox remains in indeterminate state and when clicked does not deselect all selected

Expected Behavior:
Given a grid with grouping and header select checkbox enabled to select filtered only, and grid is filtered...

1. When header select all is clicked/checked all filtered results should be checked and header select all checkbox should reflect that all rows are checked.
2. When all filtered rows are checked clicking the header select all checkbox should deselect all selected filtered rows

**Comment by makinggoodsoftware** [comment]
Hi,

Your ticket has been flagged as in_zendesk, it means that we recognize this as an issue worth investigating so we have moved it to our official support channel for customers (zendesk)

We will carry on there with the investigation and we will update you as soon as possible.

Thanks

**Comment by makinggoodsoftware** [comment]
Hi,

Thank you for your email. We are aware of this issue and have a feature request in our backlog:

You can follow the status of feature requests, bugs and releases using our pipeline: 
https://www.ag-grid.com/ag-grid-pipeline/

If you are an ag-Grid Enterprise Customer, please raise this in our Zendesk support system - contact info@ag-grid.com for access. This is our primary channel for Support.

This is the reference and summary of the relevant issue:

AG-2127
Filters - Deselect All Filtered not deselecting on edge cases

**Comment by Bosmatz** [comment]
Hello, 
Is this issue fixed?
I'm trying to deal with it right now and have no solution for how to make it works..
version: "ag-grid-community": "27.0.1", "ag-grid-enterprise": "27.0.1","ag-grid-vue3": "27.0.0"
Thanks

**Comment by Bosmatz** [comment]
@alantreadway @AndrewGlazier

**Comment by 3Dayg** [comment]
I had this same issue on v27.3, but apparently it was patched in v29.0:
Commit with the fix → https://github.com/ag-grid/ag-grid/commit/3307f95725a7f81d542a597d4494174a9923e48c

My problem, which I believe is the same as the OP’s, was as follows:

- I had checkboxSelection, groupSelectsChildren, and groupSelectsFiltered all set to true.
- Everything worked correctly (selection/deselection) until I applied a filter to the grid. After selecting a group, I wanted to deselect my current selection, but in that case, the deselection did nothing (clicking on the checkbox had no effect).

If you are experiencing this issue and cannot upgrade your library, I propose the following (somewhat sketchy) workaround:
```
{
  ...yourGridOptions,
  checkboxSelection: true,
  groupSelectsChildren: false, // set to false or omit
  groupSelectsFiltered: false, // set to false or omit
  onRowSelected: function(params) {
    if (params.node.group) {
      params.node.childrenAfterFilter.forEach((node) => {
        node.setSelected(params.node.selected, false);
      });
    }
  },
}
```

Basically, we enable checkboxSelection but disable the automatic group selection/deselection features, which are buggy when filters are applied (causing the inability to deselect). Then, we programmatically handle group node selection in the onRowSelected event.

I also suggest to check the documentation archive of Ag grid https://www.ag-grid.com/archive/27.3.0/javascript-data-grid/row-selection/#checkbox-selection , has great documentation for old versions.
